File: problems/CSES/Graph_Algorithms/road_construction/road_construction.py
import sys
# sys.setrecursionlimit(200000)

input = sys.stdin.readline

class DSU:
    def __init__(self, n):
        self.parents  = list(range(n))
        self.size = [1] * n

    def find(self, x):
        if self.parents[x] != x:
            self.parents[x] = self.find(self.parents[x]) # Path Compression

        return self.parents[x]

    def union(self, u, v):
        u = self.find(u)
        v = self.find(v)

        if u == v:
            return False

        if self.size[u] < self.size[v]:
            u, v = v, u

        self.parents[v] = u
        self.size[u] += self.size[v]

        return True

    def connected(self, a, b):
        return self.find(a) == self.find(b)


def main():
    n, m = map(int, input().split())

    dsu = DSU(n)

    components = n
    largest = 1

    for _ in range(m):
        a, b = map(int, input().split())

        a -= 1
        b -= 1

        if dsu.union(a, b):
            components -= 1

            root = dsu.find(a)
            largest = max(largest, dsu.size[root])

        print(components, largest)

main()

File: problems/CSES/Graph_Algorithms/road_construction/road_construction.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
#define pb push_back

class DSU {
private:
    vector<int> parents;
    vector<int> sz;

public:
    DSU(int n) {
        parents.resize(n);
        iota(parents.begin(), parents.end(), 0);

        sz.assign(n, 1);
    }

    int find(int x) {
        if (parents[x] != x) {
            parents[x] = find(parents[x]); // Path Compression
        }

        return parents[x];
    }

    int unite(int u, int v) {
        u = find(u);
        v = find(v);

        if (u == v) {
            return 0;
        }

        if (sz[u] < sz[v]) {
            swap(u, v);
        }

        parents[v] = u;
        sz[u] += sz[v];

        return sz[u];
    }

    bool connected(int a, int b) {
        return find(a) == find(b);
    }

    int get_size(int x) {
        return sz[find(x)];
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;

    DSU dsu(n);

    int components = n;
    int largest = 1;

    for (int i = 0; i < m; i++) {
        int a, b;
        cin >> a >> b;
        a--; b--;

        int new_sz = dsu.unite(a, b);
        if (new_sz > 0) {
            components--;
            largest = max(largest, new_sz);
        }

        cout << components << ' ' << largest << '\n';
    }

    return 0;
}
